core/src/main/java/hudson/model/Queue.java

```diff
@@ -2489,9 +2536,14 @@ public CauseOfBlockage getCauseOfBlockage() {
                     if (nodes.size() != 1)      return new BecauseLabelIsOffline(label);
                     else                        return new BecauseNodeIsOffline(nodes.iterator().next());
                 } else {
+                    if (transientCausesOfBlockage != null && label.getIdleExecutors() > 0) {
```

Comment:
This is admittedly hackish. Essentially the problem is this: we have gone through a bunch of open executor slots and checked whether the project can be built on any of them. If they all say no, we need to provide a clear explanation why. In the common case that the answer is simply that there are no executors with a matching label, we want to pick `BecauseLabelIsBusy` ~ `Queue.WaitingForNextAvailableExecutorOn`. But in general there might be multiple reasons contributing to the item staying in queue, including some much less obvious ones that the user needs to be shown.

A possible incremental improvement would be for `Node.canTake` to return a specific type for `Node.LabelMissing`, which `maintain` would then check for in the result of `JobOffer.getCauseOfBlockage` and decline to add to `reasons`.

But even that is still clumsy—we are creating a special case for the issue of label mismatches. That is bad not just for stylistic reasons; it also means that a plugin could not implement something similar to labels but subtly different, without losing some specialized behavior in `BuildableItem.getCauseOfBlockage`.

I thought of introducing some sort of numeric priority to `CauseOfBlockage` which ranks how “interesting” it is: label mismatch would be uninteresting, while lack of `Computer.BUILD` would be interesting, and a `QueueTaskDispatcher` (or `NodeProperty`) could decide how interesting its blockage is. But this may not work very well in some cases either.